storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java

```diff
@@ -0,0 +1,582 @@
+                            index.sanityCheck();
+                            return index;
+                        } catch (IOException e) {
+                            throw new KafkaException(e);
+                        }
+                    });
+
+                    TransactionIndex txnIndex = loadIndexFile(fileName, TXN_INDEX_FILE_SUFFIX, remoteLogSegmentMetadata, rlsMetadata -> {
+                        try {
+                            return remoteStorageManager.fetchIndex(rlsMetadata, IndexType.TRANSACTION);
+                        } catch (RemoteStorageException e) {
+                            throw new KafkaException(e);
+                        }
+                    }, file -> {
+                        try {
+                            TransactionIndex index = new TransactionIndex(startOffset, file);
+                            index.sanityCheck();
+                            return index;
+                        } catch (IOException e) {
+                            throw new KafkaException(e);
+                        }
+                    });
+
+                    return new Entry(offsetIndex, timeIndex, txnIndex);
+                } catch (IOException e) {
+                    throw new KafkaException(e);
+                }
+            });
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    public int lookupOffset(RemoteLogSegmentMetadata remoteLogSegmentMetadata, long offset) {
+        lock.readLock().lock();
+        try {
+            return getIndexEntry(remoteLogSegmentMetadata).lookupOffset(offset).position;
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    public int lookupTimestamp(RemoteLogSegmentMetadata remoteLogSegmentMetadata, long timestamp, long startingOffset) throws IOException {
+        lock.readLock().lock();
+        try {
+            return getIndexEntry(remoteLogSegmentMetadata).lookupTimestamp(timestamp, startingOffset).position;
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    public void close() {
+        // Make close idempotent and ensure no more reads allowed from henceforth. The in-progress reads will continue to
+        // completion (release the read lock) and then close will begin executing. Cleaner thread will immediately stop work.
+        if (!isRemoteIndexCacheClosed.getAndSet(true)) {
+            lock.writeLock().lock();
+            try {
+                log.info("Close initiated for RemoteIndexCache. Cache stats={}. Cache entries pending delete={}",
+                        internalCache.stats(), expiredIndexes.size());
+                // Initiate shutdown for cleaning thread
+                boolean shutdownRequired = cleanerThread.initiateShutdown();
+                // Close all the opened indexes to force unload mmap memory. This does not delete the index files from disk.
+                internalCache.asMap().forEach((uuid, entry) -> entry.close());
```

Comment:
Does this need to be done between `initiateShutdown` and `awaitShutdown`? Typically, this is only needed if you want to make sure sth happens during the shutdown. If not, the simpler approach is to make a single call `shutdown`.

Reply:
This is done as an optimization here for parallel processing. The `entry.close()` operation, performs close of 3 indexes per entry in the cache, hence, it closes 1024 * 3 = 3072 indexes. Closing of each index performs index resize and mmap unload which can be an expensive operation. Hence, we want to start the cleaner thread shutdown and instead of waiting for it to close, we want to continue perfomring other operations such as entry.close.

That is why we do initiate and await separately.